### ISSUE TITLE
assertion failure: more than one message downloaded?

### ISSUE THREAD

**Issue by subpardaemon** [mrts/migrate-imap-account-to-gmail#2]
during a transfer of a mail account, i stumbled upon this error many times:

`Traceback` (most recent call last):
  File "./migrate-imap-account-to-gmail.py", line 188, in <module>
    main()
  File "./migrate-imap-account-to-gmail.py", line 71, in main
    msg, flags, size, date = source_account.fetch_message(message_id)
  File "./migrate-imap-account-to-gmail.py", line 121, in fetch_message
    assert len(response) == 1
AssertionError`

len(response) is 2. i know this should not be possible (getting two answers for one message ID), but still it happens. since i'm not at home in python debugging, all i could do was to echo len(response) for each fetch, that's how i know it's 2.

i modified the script inasmuch not to ask for a confirmation (makes it hard to start the script in nohup mode), and i've changed IMAPClient so that imaplib._MAXLINE is set to 100.000 (migrating really big mailboxes).

any idea as to what causes this problem, beyond the obvious of receiving 2 responses?

**Comment by mrts** [comment]
Yes, your analysis seems correct, it looks that `IMAPClient.fetch()` returns multiple messages for the same message ID. Probably an error with the IMAP server, but you could verify whether it's duplicates of the same message or different messages.

The Python debugger is not too hard though, just `import pdb` at the top and put `if len(response) != 1: pdb.set_trace()` instead of `assert len(response) == 1` and you will get a debugger shell where you can examine the `response` structure. See https://docs.python.org/2/library/pdb.html.

But if you don't want to mess with the debugger, then the pretty-print module might help with nicer formatting of arrays, `import pprint` at the top and `if len(response) != 1: pprint.pprint(response)` instead of `assert len(response) == 1` (you should probably still exit after that by e.g. `assert False`).

Post your findings here, perhaps we can find a solution - or a workaround at least.
